### PR TITLE
[FLINK-12143] Draft

### PR DESCRIPTION
**Pull request by 1u0**

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/core/plugin/PluginConfig.java [hunk 1]
```diff
@@ -31,28 +34,37 @@
...
 	private final Optional<Path> pluginsPath;
 
-	private PluginConfig() {
-		this.pluginsPath = Optional.empty();
-	}
+	private final String[] alwaysParentFirstPatterns;
 
-	private PluginConfig(Path pluginsPath) {
-		this.pluginsPath = Optional.of(pluginsPath);
+	private PluginConfig(@Nullable Path pluginsPath, String[] alwaysParentFirstPatterns) {
```
Comment (pnowojski) [inline flink-core/src/main/java/org/apache/flink/core/plugin/PluginConfig.java:39]: You are passing `Nullable` just to wrap it with `Optional` one line later. I would either introduce `Optional` as a parameter or avoid `Nullable` and `Optional` arguments at all - this is usually a sign that method should either be  overloaded or someone should use builder.

File: flink-core/src/main/java/org/apache/flink/core/plugin/PluginLoader.java [hunk 2]
```diff
@@ -39,12 +40,12 @@
...
 
 	@VisibleForTesting
-	public PluginLoader(PluginDescriptor pluginDescriptor, ClassLoader parentClassLoader) {
+	public PluginLoader(PluginDescriptor pluginDescriptor, ClassLoader parentClassLoader, String[] alwaysParentFirstPatterns) {
 		this.pluginClassLoader =
 			new ChildFirstClassLoader(
 				pluginDescriptor.getPluginResourceURLs(),
 				parentClassLoader,
-				pluginDescriptor.getLoaderExcludePatterns());
+				ArrayUtils.concat(alwaysParentFirstPatterns, pluginDescriptor.getLoaderExcludePatterns()));
```
Comment (pnowojski) [inline flink-core/src/main/java/org/apache/flink/core/plugin/PluginLoader.java:48]: I think that the intention behind `pluginDescriptor.getLoaderExcludePatterns()` was to pass `alwaysParentFirstPatterns` (and java doc above `PluginDescriptor#loaderExcludePatterns` seems to confirm that), so one of those two is redundant?

File: flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/overlays/FlinkDistributionOverlay.java [hunk 3]
```diff
@@ -113,7 +113,7 @@ public FlinkDistributionOverlay build() {
 
 		private static File getObligatoryFileFromEnvironment(String envVariableName) {
 			String directory = System.getenv(envVariableName);
-			checkState(directory != null, "the {} environment variable must be set", envVariableName);
+			checkState(directory != null, "the %s environment variable must be set", envVariableName);
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/overlays/FlinkDistributionOverlay.java:116]: Good catch.

File: flink-dist/src/main/flink-bin/bin/pyflink-gateway-server.sh [hunk 4]
```diff
@@ -50,8 +50,8 @@ done
 log=$FLINK_LOG_DIR/flink-$FLINK_IDENT_STRING-python-$HOSTNAME.log
 log_setting=(-Dlog.file="$log" -Dlog4j.configuration=file:"$FLINK_CONF_DIR"/log4j-cli.properties -Dlogback.configurationFile=file:"$FLINK_CONF_DIR"/logback.xml)
 
-TABLE_JAR_PATH=`echo "$FLINK_ROOT_DIR"/opt/flink-table*.jar`
-PYTHON_JAR_PATH=`echo "$FLINK_ROOT_DIR"/opt/flink-python*java-binding.jar`
+TABLE_JAR_PATH=`echo "$FLINK_HOME"/opt/flink-table*.jar`
+PYTHON_JAR_PATH=`echo "$FLINK_HOME"/opt/flink-python*java-binding.jar`
```
Comment (pnowojski) [inline flink-dist/src/main/flink-bin/bin/pyflink-gateway-server.sh:54]: 🙄 this was committed after I wrote my change :/

File: flink-dist/src/main/flink-bin/bin/pyflink-stream.sh [hunk 5]
```diff
@@ -22,4 +22,4 @@ bin=`cd "$bin"; pwd`
 
 . "$bin"/config.sh
 
-"$FLINK_BIN_DIR"/flink run --class org.apache.flink.streaming.python.api.PythonStreamBinder -v "$FLINK_HOME_DIR"/opt/flink-streaming-python*.jar "$@"
+"$FLINK_BIN_DIR"/flink run --class org.apache.flink.streaming.python.api.PythonStreamBinder -v "$FLINK_HOME"/opt/flink-streaming-python*.jar "$@"
```
Comment (pnowojski) [inline flink-dist/src/main/flink-bin/bin/pyflink-stream.sh:25]: 👍

File: flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java [hunk 6]
```diff
@@ -80,7 +80,6 @@
 import java.util.Date;
```
Comment (pnowojski) [inline flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java:80]: Can you rephrase my `mid work` commit message to:

> [FLINK-12143][dist] Ship plugins in the cluster the same way as lib jars

**Review by pnowojski — CHANGES_REQUESTED** [review]
Left couple of smaller comments.

File: flink-core/src/main/java/org/apache/flink/core/plugin/PluginLoader.java (hunk 2 above)
Comment (1u0) [reply flink-core/src/main/java/org/apache/flink/core/plugin/PluginLoader.java:48]: This is what I have explicitly asked Stefan Richter about, before implementing this,
if the configuration suppose to pass it via `PluginDescriptor.getLoaderExcludePatterns()`.

The letter is designed to allow customization at individual plugin, while here we are passing a system level exclude pattern. So it was fine for him, if we merge both patterns instead of piggybacking on `PluginDescriptor.getLoaderExcludePatterns()`.

File: flink-dist/src/main/flink-bin/bin/pyflink-gateway-server.sh (hunk 4 above)
Comment (1u0) [reply flink-dist/src/main/flink-bin/bin/pyflink-gateway-server.sh:54]: "Rebase translation" based on: https://github.com/pnowojski/flink/blob/FLINK-12143/flink-dist/src/main/flink-bin/bin/pyflink-gateway-server.sh#L53
:)

File: flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java (hunk 6 above)
Comment (1u0) [reply flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java:80]: I have merged some commits together + renaming.

File: flink-dist/src/main/flink-bin/bin/pyflink-stream.sh (hunk 5 above)
Comment (1u0) [reply flink-dist/src/main/flink-bin/bin/pyflink-stream.sh:25]: This is merged now into your commit.

File: flink-core/src/main/java/org/apache/flink/core/plugin/PluginConfig.java (hunk 1 above)
Comment (1u0) [reply flink-core/src/main/java/org/apache/flink/core/plugin/PluginConfig.java:39]: Before my changes, there were two constructors, so I kind of unified them.
They say, that idiomatically java `Optional` is not supposed to be passed into constructor or  stored as field, but the use case more in the getters. In this case it would be translated to:
```java
@Nullable
private final Path pluginsPath;
...
private PluginConfig(@Nullable Path pluginsPath, String[] alwaysParentFirstPatterns) {
...
}

public Optional<Path> getPluginsPath() {
	return Optional.ofNullable(pluginsPath);
}
...
```

What do you think about such variant?
Introducing a builder now, for only two parameters, is an overkill, imo.

Comment (pnowojski) [reply flink-core/src/main/java/org/apache/flink/core/plugin/PluginConfig.java:39]: > What do you think about such variant?

You know my opinion about `@Nullable` :( +1 in general, -1 in Flink as for now.

> Introducing a builder now, for only two parameters, is an overkill, imo.

Yes, it would be an overkill. That's why I would prefer lesser evil of `Optional` in the arguments

**Comment by 1u0** [comment]
Closing it here for https://github.com/apache/flink/pull/8660.
